airflow/migrations/versions/0055_3_0_0_remove_pickled_data_from_dagrun_table.py

```diff
@@ -0,0 +1,146 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""
+remove pickled data from dagrun table.
+
+Revision ID: e39a26ac59f6
+Revises: 38770795785f
+Create Date: 2024-12-01 08:33:15.425141
+
+"""
+
+from __future__ import annotations
+
+import json
+import pickle
+from textwrap import dedent
+
+import sqlalchemy as sa
+from alembic import context, op
+from sqlalchemy import text
+from sqlalchemy.dialects import postgresql
+
+# revision identifiers, used by Alembic.
+revision = "e39a26ac59f6"
+down_revision = "38770795785f"
+branch_labels = None
+depends_on = None
+airflow_version = "3.0.0"
+
+
+def upgrade():
+    """Apply remove pickled data from dagrun table."""
+    conn = op.get_bind()
+    conf_type = sa.JSON().with_variant(postgresql.JSONB, "postgresql")
+    op.add_column("dag_run", sa.Column("conf_json", conf_type, nullable=True))
+
+    if context.is_offline_mode():
+        print(
+            dedent("""
+            ------------
+            --  WARNING: Unable to migrate the data in the 'conf' column while in offline mode!
+            --  The 'conf' column will be set to NULL in offline mode.
+            --  Avoid using offline mode if you need to retain 'conf' values.
+            ------------
+            """)
+        )
+    else:
+        BATCH_SIZE = 2
+        offset = 0
+        while True:
+            rows = conn.execute(
+                text(
+                    f"SELECT id,conf FROM dag_run WHERE conf IS not NULL order by id LIMIT {BATCH_SIZE} OFFSET {offset}"
+                )
+            ).fetchall()
+            if not rows:
+                break
+            for row in rows:
+                row_id, pickle_data = row
+
+                try:
+                    original_data = pickle.loads(pickle_data)
+                    json_data = json.dumps(original_data)
+                    conn.execute(
+                        text("""
+                                                UPDATE dag_run
+                                                SET conf_json = :json_data
+                                                WHERE id = :id
+                                            """),
+                        {"json_data": json_data, "id": row_id},
+                    )
+                except Exception as e:
+                    print(f"Error converting dagrun conf to json for dagrun ID {row_id}: {e}")
+                    continue
+            offset += BATCH_SIZE
+
+    op.drop_column("dag_run", "conf")
+
+    op.alter_column("dag_run", "conf_json", existing_type=conf_type, new_column_name="conf")
+
+
+def downgrade():
+    """Unapply Remove pickled data from dagrun table."""
+    conn = op.get_bind()
+    conf_type = sa.LargeBinary().with_variant(postgresql.BYTEA, "postgresql")
+    op.add_column("dag_run", sa.Column("conf_pickle", conf_type, nullable=True))
```

Comment:
```suggestion
    op.add_column("dag_run", sa.Column("conf", conf_type, nullable=True))
```
Should it be `conf` with `PickleType`?

Reply:
we need a different column name then `conf` as we already have `conf`. Here we are adding `conf_pickle` column and the renaming [here](https://github.com/apache/airflow/blob/fbdfd66cf4424dc3345d25e021351de98fa799cc/airflow/migrations/versions/0055_3_0_0_remove_pickled_data_from_dagrun_table.py#L146)

Reply:
What do you think about the data type? It should be PickleType, as that was what was there before. No?

Reply:
Yes its PickleType, We are setting this [here](https://github.com/apache/airflow/blob/fbdfd66cf4424dc3345d25e021351de98fa799cc/airflow/migrations/versions/0055_3_0_0_remove_pickled_data_from_dagrun_table.py#L101)

Reply:
It should be this way: https://github.com/apache/airflow/blob/2cd40ca30b1ae568e10bd3dfbfe9a77e2eba3aa5/airflow/migrations/versions/0009_1_6_0_dagrun_config.py#L41

Reply:
Yes we can do this. Done